rust/src/io/object_store.rs

```diff
@@ -54,21 +57,72 @@ impl std::fmt::Display for ObjectStore {
     }
 }
 
+const AWS_CREDS_CACHE_KEY: &str = "aws_credentials";
 #[derive(Debug)]
 struct AwsCredentialAdapter {
-    pub inner: DefaultCredentialsChain,
+    pub inner: Arc<dyn ProvideCredentials>,
+
+    // RefCell can't be shared accross threads, so we use HashMap
+    cache: Arc<RwLock<HashMap<String, Arc<aws_credential_types::Credentials>>>>,
+}
+
+impl AwsCredentialAdapter {
+    fn new(provider: Arc<dyn ProvideCredentials>) -> Self {
+        Self {
+            inner: provider,
+            cache: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
 }
 
 #[async_trait]
 impl CredentialProvider for AwsCredentialAdapter {
     type Credential = ObjectStoreAwsCredential;
+
     async fn get_credential(&self) -> ObjectStoreResult<Arc<Self::Credential>> {
-        let creds = self.inner.provide_credentials().await.unwrap();
-        Ok(Arc::new(Self::Credential {
-            key_id: creds.access_key_id().to_string(),
-            secret_key: creds.secret_access_key().to_string(),
-            token: creds.session_token().map(|s| s.to_string()),
-        }))
+        let cached_creds = {
+            let cache_value = self.cache.read().await.get(AWS_CREDS_CACHE_KEY).cloned();
```

Comment:
Should we have a `CachedCredentialProvider` to handle expiration and read-through ?

Reply:
this is okay I think. Caching should always be on when using ephemeral creds, this is the same behavior in assume role cred provider.